csrc/index_compute.cpp

```diff
@@ -3291,81 +3291,594 @@ Val* Index::eye(
+// IterDomains are the box sizes (the `boxDim` parameter in
+// `cuTensorMapEncodeTiled`).
+//
+// With this understanding, we can now define more terminologies:
+//
+// Definition 3: "boxId" is a function that takes an originating bulk IterDomain
+// as its input, and returns an IterDomain. Given an originating bulk IterDomain
+// I1, the return value, boxId(I1) is:
+//   - I1 if I1 is the inner of the split that defines it
+//   - The parent of I1 if I1 is the outer of the split that defines it
+//
+// In the above diagram, boxId(I4) is I4, and boxId(I7) is I6.
+//
+// Definition 4: "globalId" is a function that takes an originating bulk
+// IterDomain as its input, and returns another IterDomain. Given an originating
+// bulk IterDomain I1, the return value, globalId(I1) is:
+//   - The parent of I1 if I1 is the inner of the split that defines it
+//   - The grandparent of I1 if I1 is the outer of the split that defines it
+//
+// In the above diagram, globalId(I4) is I1, and globalId(I7) is I2.
+//
+// Definition 5: "innerId" is a function that takes an originating bulk who is
+// the outer of the split that defines it, and returns its sibling IterDomain.
+//
+// In the above diagram, innerId(I7) is I8.
+//
+// Definition 6: An IterDomain I1 is "global" if there exists an originating
+// bulk IterDomain I2 such that I1 == globalId(I2).
+//
+// In the above diagram, I1 and I2 are global IterDomains.
+//
+// Note that the number of global IterDomains should always be equal to the
+// number of originating bulk IterDomains, which should always be equal to the
+// dimensionality of the tensor in the eyes of TMA.
+//
+// Also note that global IterDomains can not have dependencies between each
+// other, that is, no one global IterDomains can be an ancestor or descendant of
+// another global IterDomains. That is, schedules like this is not allowed:
+//   I1 -> split -> Bulk (inner)
+//              \-> I2 (outer) -> split -> Bulk (inner)
+//                                     \-> I3 (outer)
+// Also note that the set of all global IterDomains must be "equivalent" to the
+// allocation domain of the tensor (ignoring unrelated IterTypes like broadcast
+// and reduction), as defined in `ir_utils::validateDomainEquivalence`.
+namespace {
 
-  int64_t swizzle_size = 1;
+int64_t getCpAsyncBulkTensorSwizzleSize(TensorView* smem_tv) {
   auto exprs = DependencyCheck::getAllExprsBetween(
       {smem_tv->getMaybeRFactorDomain().begin(),
        smem_tv->getMaybeRFactorDomain().end()},
       {smem_tv->getMaybeAllocationDomain().begin(),
        smem_tv->getMaybeAllocationDomain().end()});
   for (auto expr : exprs) {
     if (auto s = dynamic_cast<Swizzle*>(expr)) {
-      swizzle_size = s->inX()->extent()->evaluate().as<int64_t>();
-      break;
+      return s->inX()->extent()->evaluate().as<int64_t>();
     }
   }
+  return 1;
+}
+
+} // namespace
+
+std::pair<Val*, Val*> Index::getCpAsyncBulkGmemIndex(
```

Comment:
Could you give here an overview of the approach taken in this function? Would be helpful before diving into the function.